File: golang/d2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Printf("error opening the file: %v", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var totalScore1 uint64
	var totalScore2 uint64

	for scanner.Scan() {
		lineSplit := strings.Split(scanner.Text(), " ")
		a := actionToScore1(lineSplit[0])
		b := actionToScore1(lineSplit[1])

		totalScore1 += calScore(a, b)
		bAction := actionToScore2(a, b)
		println(a, bAction)
		totalScore2 += calScore(a, bAction)
	}

	println(totalScore1)
	println(totalScore2)

}

func actionToScore1(action string) uint64 {
	if action == "A" || action == "X" {
		return 1
	}
	if action == "B" || action == "Y" {
		return 2
	}
	if action == "C" || action == "Z" {
		return 3
	}
	return 0
}

func actionToScore2(a, b uint64) uint64 {
	// If b is 2 we need to draw so we return the same as a
	if b == 2 {
		return a
	}

	// if b is 3 we need to lose
	if b == 1 {
		switch a {
		case 1:
			return 3
		case 2:
			return 1
		default:
			return 2
		}
	}

	// At the point the only option is to win
	// if a is 1 or 2 we return the next value
	if a < 3 {
		return a + 1
	}

	// if a is 3 (max value) we return 1 that wins 3
	return 1
}

func calScore(a, b uint64) uint64 {
	if a == b {
		return b + 3
	}

	if (a == 1 && b == 3) || (a == 2 && b == 1) || (a == 3 && b == 2) {
		return b + 0
	}

	return b + 6
}

File: rust/d2/src/main.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
    let file = File::open("input.txt").expect("Fail to open file");
    let reader = BufReader::new(file);

    let mut total_score1: u32 = 0;
    let mut total_score2: u32 = 0;

    for line in reader.lines() {
        let l = line.expect("Fail to read line");

        let line_split: Vec<String> = l.split_whitespace().map(|s| s.to_string()).collect();

        let a = action_to_score1(line_split.get(0).expect("fail to get value of a"));
        let b = action_to_score1(line_split.get(1).expect("fail to get value of b"));

        total_score1 += cal_score(a, b);

        let b_action = action_to_score2(a, b);

        total_score2 += cal_score(a, b_action);
    }

    println!("{}", total_score1);
    println!("{}", total_score2);
}

fn action_to_score2(a: u32, b: u32) -> u32 {
    if b == 2 {
        return a;
    }

    if b == 1 {
        match a {
            1 => return 3,
            2 => return 1,
            _ => return 2,
        }
    }

    if a < 3 {
        return a + 1;
    }

    1
}

fn action_to_score1(action: &str) -> u32 {
    match action {
        "A" | "X" => 1,
        "B" | "Y" => 2,
        _ => 3, // C and Z
    }
}

fn cal_score(a: u32, b: u32) -> u32 {
    if a == b {
        return b + 3;
    }

    if (a == 1 && b == 3) || (a == 2 && b == 1) || (a == 3 && b == 2) {
        return b;
    }

    b + 6
}
